common/src/main/java/com/revenuecat/purchases/common/BillingAbstract.kt

```diff
@@ -0,0 +1,97 @@
+package com.revenuecat.purchases.common
+
+import android.app.Activity
+import com.android.billingclient.api.BillingClient
+import com.android.billingclient.api.BillingResult
+import com.revenuecat.purchases.ProductType
+import com.revenuecat.purchases.PurchasesError
+import com.revenuecat.purchases.models.ProductDetails
+
+typealias ProductDetailsListCallback = (List<ProductDetails>) -> Unit
+
+typealias PurchasesErrorCallback = (PurchasesError) -> Unit
+
+abstract class BillingAbstract {
+
+    @get:Synchronized
+    @set:Synchronized
+    @Volatile
+    var stateListener: StateListener? = null
+
+    @get:Synchronized
+    @Volatile
+    var purchasesUpdatedListener: PurchasesUpdatedListener? = null
+        set(value) {
+            synchronized(this@BillingAbstract) {
```

Comment:
It's not necessary, but I add it for clarity to indicate the lock should be around BillingAbstract. In this case `this` is effectively `BillingAbstract`, so it's the same to just do `synchronized(this)`.

I started adding this `this@...` after I had some synchronization issues in code similar to the following in the Purchases class:

```
 private fun getPurchasesUpdatedListener(): BillingAbstract.PurchasesUpdatedListener {
        return object : BillingAbstract.PurchasesUpdatedListener {
            override fun onPurchasesUpdated(purchases: List<@JvmSuppressWildcards PurchaseWrapper>) {
                val productChangeInProgress: Boolean
               
                synchronized(this) {
                    productChangeInProgress = state.productChangeCallback != null
                }
            }

            override fun onPurchasesFailedToUpdate(purchasesError: PurchasesError) {
                
            }
        }
    }
```

That `this` in that `synchronized(this)` refers to BillingAbstract.PurchasesUpdatedListener instead of referring to the Purchases instance. It obviously makes sense, but I decided to start always adding the @ to make sure I was synchronizing on the thing I was expecting.